### PR TITLE
Upgrade to Net 4.5.1 and enabling TransactionScopeAsyncFlow

### PR DESCRIPTION
**Pull request by danielmarbach**
This PR upgrade NServiceBus to .NET 4.5.1. This change is required for us to correctly flow the transaction scope into asynchronous operations (see [TransactionScopeAsyncFlowOption.Enabled](https://msdn.microsoft.com/en-us/library/system.transactions.transactionscopeasyncflowoption(v=vs.110).aspx))

here is what happens when you try to install the package in a project which is not targetted against .NET 4.5.1

image

And if the project is compatible

image

### CODE REVIEW

**Comment by adamralph** [comment]
Isn't this PR really 'support async-await operations inside handlers'?

The upgrade to .net 4.5.1 is a necessary change, but it's really just an implementation detail. Of course, it is important that it is publicised as a breaking change.

Ideally, I'd prefer to see the upgrade to 4.5.1 happen in one commit, and then the changes that enables in a separate commit, but this can be considered :lipstick: and shouldn't hold up the PR.

As an aside, the upgrade to 4.5.1 could even be a separate issue and PR, as an enabler, with features such as supporting async-await inside handlers marked as dependencies on it. Just an observation, not something I'm suggesting we need to do.

**Comment by danielmarbach** [comment]
Not sure where you are getting at. We can always squash and rebase when it
goes into develop. This PR is one step into the larger direction towards
async.

Am Freitag, 3. Juli 2015 schrieb Adam Ralph :

**Comment by Thorium** [comment]
+1

File: src/NServiceBus.Callbacks/NServiceBus.Callbacks.csproj [hunk 1]
```diff
@@ -9,12 +9,18 @@
...
+    <TargetFrameworkVersion>v4.5.1</TargetFrameworkVersion>
+    <RootNamespace>NServiceBus.Deploy</RootNamespace>
+    <AssemblyName>NServiceBus.Deploy</AssemblyName>
+    <TargetFrameworkVersion>v4.5.1</TargetFrameworkVersion>
     <FileAlignment>512</FileAlignment>
     <SignAssembly>true</SignAssembly>
     <AssemblyOriginatorKeyFile>..\NServiceBus.snk</AssemblyOriginatorKeyFile>
     <NuGetPackageImportStamp>b947b140</NuGetPackageImportStamp>
     <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\</SolutionDir>
+    <CreatePackages>False</CreatePackages>
```
Comment (SimonCropp) [inline src/NServiceBus.Callbacks/NServiceBus.Callbacks.csproj]: what added this?

File: src/NServiceBus.Core.Tests/App.config [hunk 2]
```diff
@@ -27,4 +27,4 @@
       <Channel Address="https://localhost/headquarter" ChannelType="Http" NumberOfWorkerThreads="3"/>
     </Channels>
   </GatewayConfig>
-</configuration>
+<startup><supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.5.1"/></startup></configuration>
```
Comment (SimonCropp) [inline src/NServiceBus.Core.Tests/App.config]: this is redundant

File: src/NServiceBus.PerformanceTests/app.config [hunk 3]
```diff
@@ -0,0 +1,3 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (SimonCropp) [inline src/NServiceBus.PerformanceTests/app.config]: delete this file

File: src/ReturnToSourceQueue/app.config [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (SimonCropp) [inline src/ReturnToSourceQueue/app.config]: delete this file

**Comment by SimonCropp** [comment]
is there no smoke test or acceptance test we should add to verify this change?

**Comment by danielmarbach** [comment]
@SimonCropp I fixed the redundant code snippets. Not sure about the smoke test or acceptance test. I could add a test which uses a `TransactionScope` without the async flow option and the show that it throws when it is disposed on another continuation. It doesn't work inside the handler because our handler infrastructure is not async yet. Not sure how much value this provides.

**Comment by danielmarbach** [comment]
Reviewed by @SimonCropp and @distantcam 

I'll merge it. @SimonCropp if you still feel strongly about the integration test ping me and I'll sort it out in another PR.

**Pull request merged**